File: static/gallery_res/demo6/demo6.cpp
/*
sum_i f_i²p_i - (sum_i f_i p_i)²

where

f_i: ℝ  
p_i: ℝ  
*/
#include <Eigen/Core>
#include <Eigen/Dense>
#include <Eigen/Sparse>
#include <iostream>
#include <set>

/**
 * demo6
 *
 * @param f  ℝ
 * @param p  ℝ
 * @return ret
 */
double demo6(
    const std::vector<double> & f,
    const std::vector<double> & p)
{
    const long _dim_0 = f.size();
    assert( f.size() == _dim_0 );
    assert( p.size() == _dim_0 );

    double _sum_0 = 0;
    for(int i=1; i<=f.size(); i++){
        _sum_0 += pow(f.at(i-1), 2) * p.at(i-1);
    }
    double _sum_1 = 0;
    for(int i=1; i<=f.size(); i++){
        _sum_1 += f.at(i-1) * p.at(i-1);
    }
    double ret = _sum_0 - pow((_sum_1), 2);
    return ret;
}


void generateRandomData(std::vector<double> & f,
    std::vector<double> & p)
{
    const int _dim_0 = rand()%10;
    f.resize(_dim_0);
    for(int i=0; i<_dim_0; i++){
        f[i] = rand() % 10;
    }
    p.resize(_dim_0);
    for(int i=0; i<_dim_0; i++){
        p[i] = rand() % 10;
    }
}


int main(int argc, char *argv[])
{
    std::vector<double> f;
    std::vector<double> p;
    generateRandomData(f, p);
    double func_value = demo6(f, p);
    std::cout<<"func_value:\n"<<func_value<<std::endl;
    return 0;
}

File: static/gallery_res/demo6/demo6.py
"""
sum_i f_i²p_i - (sum_i f_i p_i)²

where

f_i: ℝ  
p_i: ℝ  
"""
import numpy as np
import scipy
import scipy.linalg
from scipy import sparse
from scipy.integrate import quad
from scipy.optimize import minimize


def demo6(f, p):
    """
    :param :f : ℝ
    :param :p : ℝ
    """
    f = np.asarray(f)
    p = np.asarray(p)

    _dim_0 = f.shape[0]
    assert f.shape == (_dim_0,)
    assert p.shape == (_dim_0,)

    _sum_0 = 0
    for i in range(1, len(f)+1):
        _sum_0 += np.power(f[i-1], 2) * p[i-1]
    _sum_1 = 0
    for i in range(1, len(f)+1):
        _sum_1 += f[i-1] * p[i-1]
    ret = _sum_0 - np.power((_sum_1), 2)
    return ret


def generateRandomData():
    _dim_0 = np.random.randint(10)
    f = np.random.randn(_dim_0)
    p = np.random.randn(_dim_0)
    return f, p


if __name__ == '__main__':
    f, p = generateRandomData()
    print("f:", f)
    print("p:", p)
    func_value = demo6(f, p)
    print("func_value: ", func_value)